### DISCUSSION TITLE
Trying to connect two services inside GCP, need help. Setting tls programmatically.

### DISCUSSION THREAD

**Discussion by mpiorowski** [hyperium/tonic#1285] [Q&A] [answered]
So i am trying to connect two services running as Google Cloud Run.

I have a valid connection NodeJS Server -> Rust Client, but cannot make a connection between Rust Server -> Rust Client.
I am getting error:  
`Error: status: Unknown, message: "transport error", details: [], metadata: MetadataMap { headers: {} }`

So e few thing i think may be related:

1. I need to add authorization header, and right now this is a way i am doing it:
```rust
let request = Request::from_parts(
    metadata.clone(),
    Default::default(),
    UserId {
        user_id: note.user_id.to_owned(),
    },
);
```
Where metadata is:
` MetadataMap { headers: {"authorization": "Bearer xxxx" } }`

2. On nodejs server i needed to create ssl credentials, and i don't see anything related here in rust. Maybe that's the problem?

**Comment by mpiorowski** [comment]
I think i am missing tls/ssl setup, but what i did
```rust
let channel = channel_users
            .tls_config(tonic::transport::ClientTlsConfig::new())
            .context("Failed to create tls config to users service")?
            .connect()
            .await
            .context("Failed to connect to users service")?;

```
Throw me an error
```
error trying to connect: invalid peer certificate contents: invalid peer certificate: UnknownIssuer
```

**Comment by mpiorowski** [comment]
So this is how gcp do is using go:
```go
func NewConn(host string, insecure bool) (*grpc.ClientConn, error) {
        var opts []grpc.DialOption
        if host != "" {
                opts = append(opts, grpc.WithAuthority(host))
        }

        if insecure {
                opts = append(opts, grpc.WithInsecure())
        } else {
                // Note: On the Windows platform, use of x509.SystemCertPool() requires
                // go version 1.18 or higher.
                systemRoots, err := x509.SystemCertPool()
                if err != nil {
                        return nil, err
                }
                cred := credentials.NewTLS(&tls.Config{
                        RootCAs: systemRoots,
                })
                opts = append(opts, grpc.WithTransportCredentials(cred))
        }

        return grpc.Dial(host, opts...)
}
```

And i will try do something similiar using https://github.com/est31/rcgen

**Comment by mpiorowski** [comment]
Next step using rcgen, still not working :( Now i am getting: `invalid peer certificate contents: invalid peer certificate: UnknownIssuer`

```rust
        let subject_alt_names = vec![
            "xxx".to_string(),
            "localhost".to_string(),
        ];
        let cert = generate_simple_self_signed(subject_alt_names)
            .unwrap()
            .serialize_pem()
            .unwrap();
        let server_cert = cert.as_bytes();
        let tonic_cert = Certificate::from_pem(server_cert);
        let tls = ClientTlsConfig::new()
            .ca_certificate(tonic_cert)
            .domain_name("xxxx");
        let channel = channel_users
            .tls_config(tls)
            .context("Failed to create tls config to users service")?
            .connect()
            .await
            .context("Failed to connect to users service")?;
```

**Comment by mpiorowski** [comment]
This is a perfect example of what i am struggling with ;p
https://jessitron.com/2022/11/02/make-https-work-on-grpc-in-rust-load-a-root-certificate-into-the-tls-config/

**Comment by mpiorowski** [comment, marked as answer]
All i need to do was to add "tls-roots" to features........omg, such a simple solution......
And everything works with one simple line:
`let users_conn = UsersServiceClient::connect("xxxx").await?;`

**Reply by Pascualex** [reply]
Hey mate, just wanted to thank you for continuing providing context and the solution even though others didn't engage. This probably saved me hours today.

I'll add the errors I was seeing so that others may find this if they search by error code.

Before adding any features all I could see was no errors at connection at the following at request:

```rust
Status { code: Unknown, message: "h2 protocol error: http2 error: connection error detected: frame with invalid size", source: Some(tonic::transport::Error(Transport, hyper::Error(Http2, Error { kind: GoAway(b"", FRAME_SIZE_ERROR, Library) }))) }
```

After adding the `tls` feature I got a much more helpful message, at connection this time:

```rust
tonic::transport::Error(Transport, hyper::Error(Connect, HttpsUriWithoutTlsSupport(())))
```

And with `tls-roots` everything seems to be working fine.

**Reply by wspeirs** [reply]
I'm struggling with this _exact_ same issue (outlined in detail [here](https://stackoverflow.com/questions/77101427/cloud-run-grpc-via-rust-tonic)), but simply adding `tls-roots` (and `tls`) feature flags did not seem to do it for me. Do either of you have a complete working example online I could references? Thanks!

**Reply by Pascualex** [reply]
Where have you added these feature flags? They are required client-side.

**Reply by Pascualex** [reply]
To be clear I never had to [manually configure the channel](https://github.com/hyperium/tonic/discussions/1285#discussioncomment-5029149).

If I remember correctly I only had to: enable HTTP2 in the gRPC Cloud Run Service and these feature flags on the client.

**Reply by wspeirs** [reply]
I have HTTP2 enabled on the gRPC server, disabled on the client. I have `features = ["tls", "tls-roots"]` for tonic on the client, and no features specified on the server.

I'm _only_ using `DataAccessClient::connect(BACKEND_URI.as_str()).await?` to setup the client (ie, not setting up a channel), but I'm still getting the "invalid peer certificate: UnknownIssuer" problem. I'm assuming you don't have code/example on GitHub?

**Reply by Pascualex** [reply]
It's a company repo. But I can confirm I'm connecting the same way on the client. I don't see any relevant snippet on the server either.

If you are using a minimal Docker image to deploy your client server, you might be missing the certificates (not related to Axum).

I solved that for a different task prior to this, on my Dockerfile (we are using Debian):
```dockerfile
RUN apt update && apt install -y ca-certificates
```

**Reply by wspeirs** [reply]
I have that in my client Docker file as well:

```
FROM debian:12-slim

RUN apt update
RUN apt install -y ca-certificates
```

I have enabled my server to listen on the internet, and can hit it from my machine via [`grpc_health_probe`](https://github.com/grpc-ecosystem/grpc-health-probe/releases/tag/v0.4.19), but still fails during the startup of my client container. I _was_ going over a VPC, but have disabled all of that to _just_ try and get this working. Def going to make a blog post somewhere when I finally crack this :-)

**Reply by Pascualex** [reply]
Haha it's certainly very weird. I would try to run the client from my PC without Docker (enabling public access for the gRPC server) just to discard any error that can occur on the building and deployment stages.

**Reply by wspeirs** [reply]
I got it working with the gRPC server on the Internet, and adding `tls-webpki-roots` on the client side. However, I think my real issue was that `gcloud run services replace service.yaml` wasn't _actually_ doing anything, just redeploying the _same_ broken image. I notice that `println` messages I was putting in my client code weren't changing... it tough to get something working when you aren't actually changing the code :-( Thanks for the help!

**Reply by austin362667** [reply]
I think it works on these features mentioned because they handle the whole SSL handshake, while `tls` flag doesn't, meaning `tls` just handle the first certificate on chain of certificates.

**Reply by andreban** [reply]
I was running into this, but calling `connect()` on the client doesn't work because I need to add an interceptor that injects the authorization token, which needs its own channel.

What solved for me was adding the `tls-roots` feature to `tonic` , which adds `with_native_roots()` to `ClientTlsConfig`, and calling that solves the issue.

Here's how my implementation looks like:

```rust
    let token = ""; // Load your GCP token here. The `gcp_auth` create is working well for me.
    let bearer_token = format!("Bearer {}", token.as_str());
    let header_value: MetadataValue<_> = bearer_token.parse()?;

    let http_endpoint = format!("https://datastore.googleapis.com");
    let tls_config = ClientTlsConfig::new().with_native_roots();
    let channel = Channel::from_shared(http_endpoint)?
        .tls_config(tls_config)?
        .connect()
        .await?;

    let mut service = DatastoreClient::with_interceptor(channel, |mut req: Request<()>| {
        req.metadata_mut()
            .insert("authorization", header_value.clone());
        Ok(req)
    });
```
